### ISSUE TITLE
Catch boto RequestLimitExceeded and Possibly Add Retries

### ISSUE THREAD

**Issue by bflad** [jantman/awslimitchecker#187]
### Feature Description

Our production AWS accounts extensively use the AWS API and will occasionally bump up against the API request limits for various reasons. Currently awslimitchecker throws an entire stacktrace, which really needs to provide only the message. A potential solution, to at least reduce the occurrence of these and one that tools like Terraform use, is to add API retries whenever this type of error is thrown.

### Use Cases

* For heavily utilized AWS accounts
* For repeated limits monitoring that triggers administrator alerting, especially for PaaS providers such as ourselves

### Version

```
0.4.3
```

### Installation Method

```
pip install awslimitchecker # inside Jenkins Virtualenv Builder
```

### Supporting Software Versions

```
Python 2.7
```

### Actual Output

```
Traceback (most recent call last):
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/bin/awslimitchecker", line 11, in <module>
    sys.exit(console_entry_point())
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/awslimitchecker/runner.py", line 357, in console_entry_point
    r.console_entry_point()
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/awslimitchecker/runner.py", line 351, in console_entry_point
    res = self.check_thresholds()
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/awslimitchecker/runner.py", line 259, in check_thresholds
    service=self.service_name)
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/awslimitchecker/checker.py", line 417, in check_thresholds
    tmp = cls.check_thresholds()
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/awslimitchecker/services/base.py", line 274, in check_thresholds
    self.find_usage()
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/awslimitchecker/services/vpc.py", line 65, in find_usage
    self._find_usage_vpcs()
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/awslimitchecker/services/vpc.py", line 76, in _find_usage_vpcs
    vpcs = self.conn.describe_vpcs()
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/botocore/client.py", line 252, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/var/lib/jenkins/shiningpanda/jobs/105a782b/virtualenvs/59bc5cc2/local/lib/python2.7/site-packages/botocore/client.py", line 542, in _make_api_call
    raise ClientError(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (RequestLimitExceeded) when calling the DescribeVpcs operation: Request limit exceeded.
```

### Expected Output

```
# Without retries, only:
botocore.exceptions.ClientError: An error occurred (RequestLimitExceeded) when calling the DescribeVpcs operation: Request limit exceeded.
# With retries, after all attempts failed:
botocore.exceptions.ClientError: An error occurred (RequestLimitExceeded) when calling the DescribeVpcs operation: Request limit exceeded.
# With retries, after successful completion:
... normal output based on utilization ...
```

### Testing Assistance

We can certainly try helping with testing, although AWS API rate limiting is not the easiest to replicate as it has some factors outside our control (total AWS API utilization across everyone, etc.).

**Comment by jantman** [comment]
@bflad sorry for not responding sooner. I'd been under the impression that boto3 is supposed to retry internally, so I'll need to investigate why that isn't happening...

**Comment by jantman** [comment]
@bflad would it be possible for you to confirm the boto3 and botocore versions you're getting, or provide me with a ``pip freeze`` from the virtualenv?

Ok, I'm going to call it a night and try to come back to this early next week, but here's where my research/notes stand:

* The boto3 [documentation says](http://boto3.readthedocs.io/en/latest/guide/new.html#major-features) that its major features include "logic to retry failed requests". This is the first report I've gotten of awslimitchecker hitting a request limit since I switched to boto3 (and we used to get them all the time at work).
* botocore does appear to have retry logic, controlled by [data/_retry.json](https://github.com/boto/botocore/blob/develop/botocore/data/_retry.json) and the [retryhandler](https://github.com/boto/botocore/blob/develop/botocore/retryhandler.py)

I'm working on some tests for this that mock out botocore's low-level calls to URLLib, so I should be able to pretty reliably test the whole code path through boto3 and botocore...

**Comment by jantman** [comment]
@bflad I was able to reproduce getting the RequestLimitExceeded error on my personal account by running five sessions in parallel, each doing a tight loop over ``describe_vpcs()``. However, using boto3 1.2.3 and botocore 1.3.30, this request is being properly retried with a random sleep, and eventually returns the proper response without raising an exception.

The only thing that I can think at this point is that your account is so far past its request limit that even with the default of five retries, it's still not succeeding. Unfortunately botocore simply raises the error to the caller if the maximum number of retries is surpassed.

There are two paths forward that I can see:
(1) you could install awslimitchecker from the [no-suppress-logging](https://github.com/jantman/awslimitchecker/tree/no-suppress-logging) branch and then run it with ``-vv``, which will include the full boto3 and botocore debug-level logging. That would allow me to confirm whether it's actually being retried or not. Unfortunately, when the maximum number of retries is exhausted, botocore simply raises the exception; aside from debug-level logging, there's no other way to confirm that this is the case.
(2) Per [this comment](https://github.com/boto/botocore/issues/882#issuecomment-210631172), you can try copying botocore's [data/_retry.json](https://github.com/boto/botocore/blob/develop/botocore/data/_retry.json) from your virtualenv to ``~/.aws/models/_retry.json`` and change ``max_attempts`` to a higher number. Unfortunately, per https://github.com/boto/botocore/issues/882 there's currently no clearer way to do this.

However, my gut reaction to this is that if you're still seeing the exception after five retries with exponential backoff, there's something seriously wrong with requests going to your account - maybe some very badly-behaved client?

**Comment by jantman** [comment]
@bflad just a quick update that I think is relevant:

I just got a similar error this morning with awslimitchecker for my current employer's aws dev account. That's using awslimitchecker via its python API and a custom wrapper script (similar but not identical to [docs/examples/check_aws_limits.py](https://github.com/jantman/awslimitchecker/blob/master/docs/examples/check_aws_limits.py) that happens to log at info-level for everything, including boto/botocore.

```
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Resetting dropped connection: autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Resetting dropped connection: autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Resetting dropped connection: autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Resetting dropped connection: autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Resetting dropped connection: autoscaling.us-east-1.amazonaws.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Resetting dropped connection: autoscaling.us-east-1.amazonaws.com
Checking AWS resource usage; WARNING threshold 80% of limit, CRITICAL threshold 99% of limit
Traceback (most recent call last):
  File "./check_aws_limits.py", line 289, in <module>
    checker.run(error_on_warning=args.error_on_warning, verbose=args.verbose)
  File "./check_aws_limits.py", line 249, in run
    warnings, criticals = self.run_for_regions(regions, verbose=verbose)
  File "./check_aws_limits.py", line 230, in run_for_regions
    w, c = self.run_for_region(region=region, verbose=verbose)
  File "./check_aws_limits.py", line 201, in run_for_region
    return self.check_limits(region, verbose=verbose)
  File "./check_aws_limits.py", line 141, in check_limits
    checker.check_thresholds()
  File "/home/ec2-user/jenkins/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/lib/python2.7/site-packages/awslimitchecker/checker.py", line 417, in check_thresholds
    tmp = cls.check_thresholds()
  File "/home/ec2-user/jenkins/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/lib/python2.7/site-packages/awslimitchecker/services/base.py", line 274, in check_thresholds
    self.find_usage()
  File "/home/ec2-user/jenkins/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/lib/python2.7/site-packages/awslimitchecker/services/autoscaling.py", line 72, in find_usage
    alc_marker_param='NextToken'
  File "/home/ec2-user/jenkins/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/lib/python2.7/site-packages/awslimitchecker/utils.py", line 159, in paginate_dict
    result = function_ref(*argv, **pass_kwargs)
  File "/home/ec2-user/jenkins/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/lib/python2.7/site-packages/botocore/client.py", line 278, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/ec2-user/jenkins/workspace/ReleaseEngineering/awslimitchecker/awslimitchecker_dev/lib/python2.7/site-packages/botocore/client.py", line 572, in _make_api_call
    raise ClientError(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (Throttling) when calling the DescribeAutoScalingGroups operation: Rate exceeded
```

Those info-level botocore logs confirm that it's being retried, but the maximum number of retries is exceeded and the original exception is raised by botocore.

There's been some discussion on https://github.com/boto/botocore/issues/926 which I opened to try and expose this information to code using boto, so I'm going to try and cut a pull request for that and then modify awslimitchecker to take it into account.

**Comment by jantman** [comment]
I've opened https://github.com/boto/botocore/pull/965 against botocore to expose retry information in exceptions. Once that's merged and released, I can at least pass this on in limit checker, if not have an option to wait a while and try again.

**Comment by jantman** [comment]
This should now be tracking https://github.com/boto/botocore/pull/1024, which should be merged soon. I'll come back to this once that is in a botocore release.

To be clear, the upstream change will be to include information about whether the maximum number of retries were reached, and if so, how many that was, in the exception message.

**Comment by jantman** [comment]
Ok, this was merged to botocore master in https://github.com/boto/botocore/commit/c3babf30df041ff064882f5779b96e1eeb1e4886 and released last night as [1.4.51](https://github.com/boto/botocore/blob/1.4.51/CHANGELOG.rst).

At the very least, we should get more-helpful exception messages that clearly show when requests have been retried but reached the maximum number of retries.

Beyond that, I'm not sure what I can really do. I'm going to close this issue; please see my [comment above](https://github.com/jantman/awslimitchecker/issues/187#issuecomment-220777272) for some possible solutions.
